tecken/storage.py

```diff
@@ -113,15 +115,31 @@ def __repr__(self):
             f"backend={self.backend!r}>"
         )
 
-    def get_storage_client(self, **config_params):
-        """Return a backend-specific client, overriding default config parameters.
+    def get_storage_client(self):
+        """Return a backend-specific client with timeouts suitable for uploads.
 
         TODO(jwhitlock): Build up StorageBucket API so users don't work directly with
         the backend-specific clients (bug 1564452).
         """
-        return get_storage_client(
-            endpoint_url=self.endpoint_url, region_name=self.region, **config_params
-        )
+        if not hasattr(self.clients, "storage"):
+            self.clients.storage = get_storage_client(
+                endpoint_url=self.endpoint_url,
+                region_name=self.region,
+                read_timeout=settings.S3_PUT_READ_TIMEOUT,
+                connect_timeout=settings.S3_PUT_CONNECT_TIMEOUT,
+            )
+        return self.clients.storage
+
+    def get_lookup_client(self):
```

Comment:
I'm not sure I understand your comment and I can't tell if we're communicating past each other and I want to double-check my thinking, so I'll go through a longer explanation.

Tecken is connecting to AWS S3 to check for the existence of symbols files in S3 (lookup) and to upload files to S3 (put). Currently and in this PR, it uses two separate clients with different connect and read timeouts.

The settings values are:

```
S3_LOOKUP_CONNECT_TIMEOUT = _config(
    "S3_LOOKUP_CONNECT_TIMEOUT",
    default="2",
    parser=int,
    doc="S3 lookup connection timeout in seconds.",
)
S3_LOOKUP_READ_TIMEOUT = _config(
    "S3_LOOKUP_READ_TIMEOUT",
    default="4",
    parser=int,
    doc="S3 lookup read timeout in seconds.",
)

# When testing S3 PUT in Stage, the longest PUTs take 20 seconds.
S3_PUT_CONNECT_TIMEOUT = _config(
    "S3_PUT_CONNECT_TIMEOUT",
    default="10",
    parser=int,
    doc="S3 put connection timeout for uploads in seconds.",
)
# If upload takes longer than this it's probably best to back off.
# The client will likely get a 504 error and will retry soon again.
S3_PUT_READ_TIMEOUT = _config(
    "S3_PUT_READ_TIMEOUT",
    default="30",
    parser=int,
    doc="S3 put read timeout in seconds.",
)
```

The comments for the settings read to me like the author thinks the read timeout applies to the entire request--not the tiny period of time in the read loop to wait for the next available byte.

Ultimately, we want Tecken to successfully look up or download a file in the shortest period of time as much as possible.

In success scenarios, these timeout values make no difference. It's the failure scenarios we need to be thinking about.

When a timeout happens, botocore will retry the request. If I'm reading the docs correctly, it'll retry the S3 requests 4 times.

In a failure scenario where it can't connect, it's likely  either a major AWS S3 service outage in which case it doesn't matter how long Tecken waits or it's an ephemeral network something-or-other. Service outages do happen, but I don't think any of our timeout values help with a service outage. I think we want to set the timeout such that the code fails fast and then retries to account for periodic ephemeral issues. 10 seconds is too big a value for that.

In a failure scenario where it times out waiting for the next byte, the thinking is roughly the same--it's likely either a major AWS S3 service outage or it's an ephemeral network something-or-other. Timeout values probably won't help with a service outage, but do help with ephemeral issues. We want the code to fail fast and retry. 20 seconds is way too big a value for that. The "the longest PUTs take 20 seconds" comment makes me think the author thought the read timeout applied to the whole request.

Thus, I think the current code is based on a misunderstanding and the timeouts are excessive such that they unhelpfully increase the amount of time it takes in failure situations that are probably not uncommon.

I think we want to unify the clients and set the timeouts to something reasonable. If it were me, I'd probably have a single client and set all the timeouts to 5 which is what we use for requests (the python library) requests in `tecken/librequests.py`.

I mentioned this issue in this PR because maybe it's worth removing the two-client approach and reducing the complexity in this PR. However, the client/timeout code has been this way for a while so we could let it slide to a follow-up PR.

Reply:
I talked with @smarnach and I had missed that the read timeout probably doesn't even pertain to PUTs. That clarifies what @smarnach was saying in his comment to me.